### PR TITLE
Hostage transfers

### PR DESCRIPTION
**Pull request by colorlace**
This pull request introduces hostages into the backend. 
- selecting hostages
- transfering hostages to the other room
- wiping hostages when a leader changes or a new round starts

### CODE REVIEW

**Comment by colorlace** [comment]
I just realized that hostages need to transfer when they've all been selected and there is no next round. That's the last transfer before the winners and losers are announced!

**Comment by colorlace** [comment]
I notice that none of the things I've done immediately affect the users' UIs. This is where knowing what's going on with those channels and websockets helps.

File: .ruby-version [hunk 1]
```diff
@@ -1 +1 @@
-ruby-2.6.5
+2.7.2
```
Comment (efatsi) [inline .ruby-version:1]: derrrrp, sorry i forgot to update this.

File: app/models/actions/select_hostage.rb [hunk 2]
```diff
@@ -3,6 +3,17 @@ def resolve
     to_players.update(is_hostage: true)
   end
 
+  def self.allowed_by?(player)
+    game = player.game
+    #binding.pry
```
Comment (efatsi) [inline app/models/actions/select_hostage.rb:8]: 🧹

File: app/models/actions/start_next_round.rb [hunk 3]
```diff
@@ -5,15 +5,20 @@ def self.immediate?
...
   def self.allowed_by?(player)
     game = player.game
-
     [
       game.hosted_by?(player.user),
-      game.current_round.nil? || game.current_round.over?,
-      game.next_round.present?
+      game.current_round.nil?,
+      game.next_round.present?,
+      game.last_round.nil? || game.last_round.all_hostages_selected
```
Comment (efatsi) [inline app/models/actions/start_next_round.rb:12]: good stuff right here

File: spec/factories/action_factory.rb [hunk 4]
```diff
@@ -26,5 +26,9 @@
         type { klass.name }
       end
     end
+
+    trait :start_next_round do
+      type { "Actions::StartNextRound" }
+    end
```
Comment (efatsi) [inline spec/factories/action_factory.rb:32]: this is actually unnecessary thanks to the block immediately above it that meta programs all these types out

File: spec/factories/game_factory.rb [hunk 5]
```diff
@@ -3,9 +3,31 @@
...
       after(:create) do |game|
         create(:round, :active, game: game, number: 1)
+        create(:round, status: "waiting", game: game, number: 2)
+        create(:round, status: "waiting", game: game, number: 3)
+        create(:room, game: game)
+        create(:room, game: game)
+      end
+    end
+
+    trait :in_first_round do
```
Comment (efatsi) [inline spec/factories/game_factory.rb:26]: did you mean to name this `after_first_round`?

File: spec/models/room_spec.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      let!(:player3) { create(:player, game: game, room: subject, is_hostage: true) }
+      let!(:player4) { create(:player, game: game, room: subject, is_hostage: false) }
+
+
+      it "all_hostages_selected is true" do
+        expect(subject.all_hostages_selected).to eq(true)
+      end
+
+      context "but then time runs out and the round ends" do
+        before { game.rounds.first.update(status: "active", started_at: 61.seconds.ago, duration: 60) }
```
Comment (efatsi) [inline spec/models/room_spec.rb:24]: should this be `status: "done"`?

File: spec/models/room_spec.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      let!(:player1) { create(:player, game: game, room: subject, is_hostage: true) }
+      let!(:player2) { create(:player, game: game, room: subject, is_hostage: true) }
+      let!(:player3) { create(:player, game: game, room: subject, is_hostage: false) }
+
+      it "all_hostages_selected is false" do
+        expect(subject.all_hostages_selected).to eq(false)
+      end
+    end
+  end
+end
```
Comment (efatsi) [inline spec/models/room_spec.rb:54]: A+ spec file here

**Review by efatsi — APPROVED** [review]
2 real questions, and then a bunch of spammy / praise comments

**Comment by efatsi** [comment]
Merging this into `main`

**Pull request closed without merging**